### PR TITLE
Tests for runtime field queries with fbf aggs

### PR DESCRIPTION
**Pull request by nik9000**
This adds a few tests for runtime field queries applied to
"filter-by-filter" style aggregations. We expect to still be able to
use filter-by-filter aggregations to speed up collection when the top
level query is a runtime field. You'd think that filter-by-filter would
be slow when the top level query is slow, like it is with runtime
fields, but we only run filter-by-filter when we can translate each
aggregation bucket into a quick query. So long as the results of those
queries don't "overlap" we shouldn't end up running the slower top level
query more times than we would during regular collection.

This also adds some javadoc to that effect to the two places where we
chose between filter-by-filter and a "native" aggregation
implementation.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregatorTests.java [hunk 1]
```diff
@@ -407,6 +415,44 @@ public void testOverlappingRanges() throws IOException {
...
+        long totalDocs = (long) RangeAggregator.DOCS_PER_RANGE_TO_USE_FILTERS * 4;
+        SearchLookup lookup = new SearchLookup(s -> null, (ft, l) -> null);
+        StringFieldScript.LeafFactory scriptFactory = ctx -> new StringFieldScript("dummy", Map.of(), lookup, ctx) {
+            @Override
+            public void execute() {
+                emit("cat");
+            }
+        };
+        Query query = new StringScriptFieldTermQuery(new Script("dummy"), scriptFactory, "dummy", "cat", false);
+        debugTestCase(new RangeAggregationBuilder("r").field(NUMBER_FIELD_NAME).addRange(0, 1).addRange(1, 2).addRange(2, 3), query, iw -> {
```
Comment (not-napoleon) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregatorTests.java:436]: Nit - I think the formatting here isn't following the new standard; Can you just run the auto-formatter on this, please?

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 2]
```diff
@@ -536,6 +538,62 @@ protected void doAssertReducedMultiBucketConsumer(Aggregation agg, MultiBucketCo
         }
     }
 
+    /**
+     * Execute and aggregation and collect its {@link Aggregator#collectDebugInfo debug}
+     * information. Unlike {@link #testCase} this doesn't randomly create an
+     * {@link Aggregator} per leaf and perform partial reductions. It always
+     * creates a single {@link Aggregator} so we can get consistent debug info.
+     */
+    protected <R extends InternalAggregation> void debugTestCase(
```
Comment (not-napoleon) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:547]: Thanks for adding this!

**Review by not-napoleon — APPROVED** [review]
LGTM.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregatorTests.java (hunk 1 above)
Comment (nik9000) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregatorTests.java:436]: Surprisingly, this is what the standard looks like. If I had to guess the lambdas are letting this bunch up somehow.

Comment (not-napoleon) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregatorTests.java:436]: I stand corrected.  Thanks for checking!

**Pull request merged**